gflownet/envs/crystals/crystal.py

```diff
@@ -0,0 +1,488 @@
+        # all the methods of that environment, but it will have to be reinitialized using
+        # proper lattice system from space group once that is determined.
+        # Triclinic was used because it doesn't force any initial starting angles.
+        self.lattice_parameters = LatticeParameters(
+            lattice_system=TRICLINIC, **self.lattice_parameters_kwargs
+        )
+
+        # 0-th element of state encodes current stage: 0 for composition,
+        # 1 for space group, 2 for lattice parameters
+        self.source = (
+            [0]
+            + self.composition.source
+            + self.space_group.source
+            + self.lattice_parameters.source
+        )
+
+        # start and end indices of individual substates
+        self.composition_state_start = 1
+        self.composition_state_end = self.composition_state_start + len(
+            self.composition.source
+        )
+        self.space_group_state_start = self.composition_state_end
+        self.space_group_state_end = self.space_group_state_start + len(
+            self.space_group.source
+        )
+        self.lattice_parameters_state_start = self.space_group_state_end
+        self.lattice_parameters_state_end = self.lattice_parameters_state_start + len(
+            self.lattice_parameters.source
+        )
+
+        # start and end indices of individual submasks
+        self.composition_mask_start = 0
+        self.composition_mask_end = self.composition_mask_start + len(
+            self.composition.action_space
+        )
+        self.space_group_mask_start = self.composition_mask_end
+        self.space_group_mask_end = self.space_group_mask_start + len(
+            self.space_group.action_space
+        )
+        self.lattice_parameters_mask_start = self.space_group_mask_end
+        self.lattice_parameters_mask_end = self.lattice_parameters_mask_start + len(
+            self.lattice_parameters.action_space
+        )
+
+        self.eos = self.lattice_parameters.eos
+        self.stage = Stage.COMPOSITION
+        self.max_action_length = max(
+            max(len(a) for a in self.composition.action_space),
+            max(len(a) for a in self.space_group.action_space),
+            max(len(a) for a in self.lattice_parameters.action_space),
+        )
+        self.done = False
+
+        # Conversions
+        self.state2proxy = self.state2oracle
+        self.statebatch2proxy = self.statebatch2oracle
+        self.statetorch2proxy = self.statetorch2oracle
+
+        super().__init__(**kwargs)
+
+    def _set_lattice_parameters(self):
+        """
+        Sets LatticeParameters conditioned on the lattice system derived from the SpaceGroup.
+        """
+        crystal_system = self.space_group.get_crystal_system()
+
+        if crystal_system is None:
+            raise ValueError(
+                "Cannot set lattice parameters without crystal system determined in the space group."
+            )
```

Comment:
I think this is incorrect for a couple of reasons:
- 1. get_crystal_system is meant to return the name of the crystal system. Actually, it will never return None but instead it will return the string "None" if the crystal system has not been set. 
- 2. I think what we want here is anyway not the crystal system but the crystal lattice system, don't we? That would be:
```suggestion
        if self.space_group.state[0] == 0:
            raise ValueError(
                "Cannot set lattice parameters without crystal system determined in the space group."
            )
```

Reply:
2. No, the lattice system and the crystal system are set simultaneously because they are both set via the crystal-lattice system, which is the first dimension of the space group env.

Reply:
Note that I have made an edit of the error message too